File: calculation_multi_thread.c
#include <pthread.h>
#include <stdio.h>
#include <time.h>
#include "calculation_thread.h"

#define MAX_CALCULATIONS 3

int main(){
    clock_t start_time = clock();

    pthread_t threads[MAX_CALCULATIONS];
    Calculation calculations[MAX_CALCULATIONS];
    float *results[MAX_CALCULATIONS];

    threads[0] = 0;
    calculations[0].a = 2;
    calculations[0].b = 6;
    calculations[0].operation = '*';

    threads[1] = 1;
    calculations[1].a = 1;
    calculations[1].b = 4;
    calculations[1].operation = '+';

    threads[2] = 2;
    calculations[2].a = 5;
    calculations[2].b = 2;
    calculations[2].operation = '-';

    for(int i = 0; i < MAX_CALCULATIONS; i++){
        pthread_create(&threads[i], NULL, calculate_thread, &calculations[i]);
    }

    for(int i = 0; i < MAX_CALCULATIONS; i++){
        pthread_join(threads[i], (void **)&results[i]);
    }

    float final_result = *results[0] + *results[1] * *results[2];
    printf("Result: %.2f\n", final_result); 

    clock_t finish_time = clock();
    double execution_time = (double)(finish_time - start_time) / CLOCKS_PER_SEC;
    printf("Execution time: %.6lf seconds\n", execution_time);
    
    return 0;
}

File: calculation_multi_thread.py
import threading
import time

# define class Calculation (like the Calculation struct)
class Calculation:
    def __init__(self, a, b, operation):
        self.a = a
        self.b = b
        self.operation = operation

def calculate_thread(calculation, result_list, index):
    if calculation.operation == '+':
        result_list[index] = calculation.a + calculation.b
    elif calculation.operation == '-':
        result_list[index] = calculation.a - calculation.b
    elif calculation.operation == '*':
        result_list[index] = calculation.a * calculation.b
    elif calculation.operation == '/':
        result_list[index] = calculation.a / calculation.b
    else:
        result_list[index] = None

MAX_CALCULATIONS = 3

# list of all calculation datas
calculations = [
    Calculation(2, 6, '*'),
    Calculation(1, 4, '+'),
    Calculation(5, 2, '-')
]

# list containing results
results = [None] * MAX_CALCULATIONS

# list containing threads
threads = []

start_time = time.time()

# create and start thread
for i in range(MAX_CALCULATIONS):
    thread = threading.Thread(target=calculate_thread, args=(calculations[i], results, i))
    threads.append(thread)
    thread.start()

# thread join
for thread in threads:
    thread.join()

# final result calculation
final_result = results[0] + results[1] * results[2]
print(f"Result: {final_result:.2f}")

finish_time = time.time()
execution_time = finish_time - start_time
print(f"Execution time: {execution_time:.12f}")